mlflow/sklearn/__init__.py

```diff
@@ -455,6 +460,323 @@ def load_model(model_uri):
+        registered_dataset_list = self._eval_dataset_info_map[run_id][eval_dataset_name]
+
+        index = len(registered_dataset_list)
+        for i, id_i in enumerate(registered_dataset_list):
+            if eval_dataset_id == id_i:
+                index = i
+                break
+
+        if index == len(registered_dataset_list):
+            # register new eval dataset
+            registered_dataset_list.append(eval_dataset_id)
+
+        return self.gen_name_with_index(eval_dataset_name, index)
+
+    def register_prediction_result(self, model, eval_dataset_name, predict_result):
+        """
+        In `patched_predict`, register the prediction result instance with the model id and eval
+        dataset name. e.g.
+
+        ```
+        prediction_result = model_1.predict(eval_X)
+        ```
+        then we need register an item into the `pred_result_id_to_dataset_name_and_run_id` map like
+        below:
+        id(prediction_result) --> (inspected_original_var_name_of(eval_X), id(model_1))
+
+        With this map `dataset_id_to_dataset_name_and_run_id`, in following patched metric API,
+        we can query the eval dataset name and the model id via the `y_pred` argument instance.
+        """
+        value = (eval_dataset_name, self.get_run_id_for_model(model))
+        prediction_result_id = id(predict_result)
+        self._pred_result_id_to_dataset_name_and_run_id[prediction_result_id] = value
+
+        def clean_id(id_):
+            _get_autolog_training_status()._pred_result_id_to_dataset_name_and_run_id \
+                .pop(id_, None)
+
+        # When the `predict_result` object being GCed, its ID may be reused, so register a finalizer
+        # to clear the ID from the dict for preventing wrong ID mapping.
+        weakref.finalize(predict_result, clean_id, prediction_result_id)
+
+    @staticmethod
+    def gen_metric_call_command(self_obj, metric_fn, *call_pos_args, **call_kwargs):
+        """
+        Generate metric function call command string like `metric_fn(arg1, arg2, ...)`
+        Note: this method include inspecting argument variable name.
+         So should be called directly from the "patched method", to ensure it capture
+         correct argument variable name.
+        """
+
+        arg_list = []
+
+        def arg_to_str(arg):
+            if arg is None or np.isscalar(arg):
+                if isinstance(arg, str) and len(arg) > 32:
+                    # truncate too long string
+                    return repr(arg[:32] + '...')
+                return repr(arg)
+            else:
+                # dataset arguments or other non-scalar type argument
+                return _inspect_original_var_name(arg, fallback_name=f'<{arg.__class__.__name__}>')
+
+        param_sig = inspect.signature(metric_fn).parameters
+        param_sig_keys = list(param_sig.keys())
+
+        if self_obj is not None:
+            param_sig_keys.pop(0)
+
+        param_sig_keys = [
```

Comment:
Can we add a small inline comment explaining why we perform this filtering? I.e. to get rid of `*args` and `**kwargs`.